src/Compiler/Driver/ReuseTcResults/TcResultsPickle.fs

```diff
@@ -0,0 +1,1745 @@
+        p_tup2
+            p_ucref
+            p_tys_new
+            (uc, tinst)
+            st
+
+    | TType_anon (anonInfo, l) ->
+         p_byte 7 st
+         p_tup2
+             p_anonInfo
+             p_tys_new
+             (anonInfo, l) 
+             st
+
+and p_tys_new l = 
+    let _count = l.Length
+    p_list p_ty_new l
+
+and p_expr_new (expr: Expr) st =
+    match expr with
+    | Expr.Link e -> p_byte 0 st; p_expr_new e.Value st
+    | Expr.Const (x, m, ty)              -> p_byte 1 st; p_tup3 p_const p_dummy_range p_ty_new (x, m, ty) st
+    | Expr.Val (a, b, m)                 -> 
+        p_byte 2 st
+        p_tup4 
+            p_vref_new 
+            p_vrefFlags 
+            p_dummy_range 
+            (p_non_null_slot p_Val_new)
+            (a, b, m, a.binding) 
+            st
+    | Expr.Op (a, b, c, d)                 -> p_byte 3 st; p_tup4 p_op_new  p_tys_new p_exprs_new p_dummy_range (a, b, c, d) st
+    | Expr.Sequential (a, b, c, d)      -> p_byte 4 st; p_tup4 p_expr_new p_expr_new p_int p_dummy_range (a, b, (match c with NormalSeq -> 0 | ThenDoSeq -> 1), d) st
+    | Expr.Lambda (_, a1, b0, b1, c, d, e)   -> p_byte 5 st; p_tup6 (p_option p_Val) (p_option p_Val) p_Vals p_expr_new p_dummy_range p_ty_new (a1, b0, b1, c, d, e) st
+    | Expr.TyLambda (_, b, c, d, e)        -> p_byte 6 st; p_tup4 p_tyar_specs_new p_expr_new p_dummy_range p_ty_new (b, c, d, e) st
+    | Expr.App (funcExpr, formalType, typeArgs, args, range)           -> 
+        p_byte 7 st
+        
+        p_expr_new funcExpr st
+        p_ty_new formalType st
+        p_tys_new typeArgs st
+        p_exprs_new args st
+        p_dummy_range range st
+    | Expr.LetRec (a, b, c, _)            -> p_byte 8 st; p_tup3 p_binds p_expr_new p_dummy_range (a, b, c) st
+    | Expr.Let (a, b, c, _)               -> p_byte 9 st; p_tup3 p_bind p_expr_new p_dummy_range (a, b, c) st
+    | Expr.Match (_, a, b, c, d, e)         -> p_byte 10 st; p_tup5 p_dummy_range p_dtree p_targets p_dummy_range p_ty_new (a, b, c, d, e) st
+    | Expr.Obj (_, b, c, d, e, f, g)       -> p_byte 11 st; p_tup6 p_ty_new (p_option p_Val) p_expr_new p_methods p_intfs p_dummy_range (b, c, d, e, f, g) st
+    | Expr.StaticOptimization (a, b, c, d) -> p_byte 12 st; p_tup4 p_constraints p_expr_new p_expr_new p_dummy_range (a, b, c, d) st
+    | Expr.TyChoose (a, b, c)            -> p_byte 13 st; p_tup3 p_tyar_specs_new p_expr_new p_dummy_range (a, b, c) st
+    | Expr.Quote (ast, _, _, m, ty)         -> p_byte 14 st; p_tup3 p_expr_new p_dummy_range p_ty_new (ast, m, ty) st
+    | Expr.WitnessArg (traitInfo, m) -> p_byte 15 st; p_trait traitInfo st; p_dummy_range m st
+    | Expr.DebugPoint (_, innerExpr) -> 
+        p_byte 16 st
+        p_expr_new innerExpr st
+    
+and p_exprs_new = p_list p_expr_new
+
+and p_ucref_new (UnionCaseRef(a, b)) st =
+    p_tup3
+        (p_tcref "ucref") 
+        p_string 
+        (p_non_null_slot p_entity_spec_new)
+        (a, b, a.binding) 
```

Comment:
As noted earlier, bindings should not be pickled separately... instead, the type references should be pickled in a way that they contain all the necessary information to auto-restore the bindings.